Ensure ShibbolethRestController only redirects to trusted URLs after authentication

tdonohue opened the pull request:
This bug was discovered/reported by LGTM in the [ShibbolethRestController](https://lgtm.com/projects/g/DSpace/DSpace/snapshot/8276a236d2f44297651b311b0be1c76a89b5d05b/files/dspace-server-webapp/src/main/java/org/dspace/app/rest/ShibbolethRestController.java?sort=name&dir=ASC&mode=heatmap#xa5d3dfed284111b7:1)

## Description
The `ShibbolethRestController` will allow redirection to arbitrary URLs after successful authentication.  This is a security issue if individuals are fooled into clicking the below link while logged in.

This behavior can even be reproduced on the demo REST API by doing the following:
1. Login the the REST demo site: https://dspace7.4science.cloud/server/login.html
2. Then, visit this link:  https://dspace7.4science.cloud/server/api/authn/shibboleth?redirectUrl=http://google.com
3. You will be redirected to the Google homepage

In this PR, I've modified this Controller to ONLY allow redirects to URLs that have the same hostname as either `dspace.server.url` or `dspace.ui.url` configurations.  This should allow Shibboleth to work for any REST API or UI path, but not allow redirects to arbitrary locations.

## Instructions for Reviewers
* Review code & verify the new ITs which prove the functionality.  
* Optionally, retest above example on a local REST API, and verify that passing "http://google.com" now returns a `400 Bad Request` response.
* Ideally, someone with Shibboleth can test this new setup to verify that it works as expected with Shibboleth enabled for DSpace.

tdonohue committed:
Ensure ShibbolethRestController only redirects to trusted URLs

paulo-graca commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/ShibbolethRestController.java:
```diff
@@ -53,8 +55,20 @@ public void shibboleth(HttpServletResponse response,
+        String redirectHostName = Utils.getHostName(redirectUrl);
+        String serverHostName = Utils.getHostName(configurationService.getProperty("dspace.server.url"));
+        String clientHostName = Utils.getHostName(configurationService.getProperty("dspace.ui.url"));
+        if (StringUtils.equalsAnyIgnoreCase(redirectHostName, serverHostName, clientHostName)) {
```
@tdonohue since we are in shibboleth context, this is what I did in the current Shibboleth SLO PR:

I've added a specific method to that particular authentication method that retrieves a validated URL. To only allows redireccion for what is only configured.
```

    public static String getURLFromLogoutActionRequest(HttpServletRequest request) throws ShibAuthenticationException {
        String returnURL = null;
        // verify if we have shibboleth parameters (action and return)
        String action = request.getParameter("action");
        String shibLogoutURL = request.getParameter("return");

        // is shibboleth action for logout?
        if (ShibAuthentication.SHIBBOLETH_LOGOUT_ACTION.equals(action) && StringUtils.isNotBlank(shibLogoutURL)) {
            // for security issues we still need to validate return param
            ConfigurationService configurationService = DSpaceServicesFactory.getInstance().getConfigurationService();
            String serverUrl = Utils.getBaseUrl(configurationService.getProperty("dspace.server.url"));
            serverUrl = getSecureURL(serverUrl, request);
            if (!shibLogoutURL.startsWith(serverUrl)) {
                throw new ShibAuthenticationException("Invalid 'return' param");
            }
            returnURL = shibLogoutURL;
        }
        return returnURL;
    }

```
Meanwhile, I did some refactoring and because of higher priority stuff and I bug I found with logout I didn't got the chance to rebase my work. 

This work it was based (I think) on @atarix83 work with Shibboleth authentication PR. He added a methods I think it was Utils.getBaseUrl to retrieve the correct and absolute URL. I think, since we are at Shibboleth context it makes sense to try to use those methods to validate the URL redirection.

tdonohue replied:
@paulo-graca : I'm not seeing much difference between your method and mine?  In mine, I'm doing a Hostname comparison to ensure the `redirectUrl` param matches the Hostname of either `dspace.server.url` or `dspace.ui.url`.  In your code pasted above, it looks like you are just making sure the `redirectUrl` *starts with* the same base URL as `dspace.server.url` (which is almost the same as a Hostname, without the HTTP/HTTPS).

So, I'm not sure how to respond here.  Did you want me to make changes to the code I've added to `ShibbolethRestController` to no longer validate against the Hostname, but check more of the URL?

paulo-graca reviewed (commented):
I'm almost mostly sure this PR will work in my case because shibboleth and server are in the same place, but not so sure if they are in different place. A shibboleth configuration will tell where to redirected and this validation doesn't consider that.

On my work with SLO I also did this, but differently. Perhaps a more complex solution :) I've pasted in a comment bellow.

I will also test this PR and give feedback.

paulo-graca commented:
I've just tested and it works as expected on both sides (HAL browser and Angular).

paulo-graca reviewed (approved):
No, just to say I did this same thing but differently, perhaps in a complex way :)
The only thing I pointed out was I used Utils.getBaseUrl that was added in the Shibboleth PR.

Since it works for me, I will give my thumbs up! Thank you @tdonohue !

tdonohue committed:
Add a SuppressWarnings annotation for LGTM as it doesn't detect that we fixed this.

tdonohue commented:
Merging with one approval (NOTE: I added a final commit here just to add an `@SuppressWarnings` annotation to let LGTM know we fixed this bug.  It seems like LGTM didn't automatically detect the fix in this case)
